library/alloc/src/rc.rs

```diff
@@ -1892,15 +1892,17 @@ impl<T> Weak<T> {
     pub fn as_ptr(&self) -> *const T {
         let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);
 
-        // SAFETY: we must offset the pointer manually, and said pointer may be
-        // a dangling weak (usize::MAX) if T is sized. data_offset is safe to call,
-        // because we know that a pointer to unsized T was derived from a real
-        // unsized T, as dangling weaks are only created for sized T. wrapping_offset
-        // is used so that we can use the same code path for the non-dangling
-        // unsized case and the potentially dangling sized case.
-        unsafe {
-            let offset = data_offset(ptr as *mut T);
-            set_data_ptr(ptr as *mut T, (ptr as *mut u8).wrapping_offset(offset))
+        if is_dangling(self.ptr) {
+            // If the pointer is dangling, we return a null pointer as the dangling sentinel.
+            // We can't return the usize::MAX sentinel, as that could valid if T is ZST.
+            // SAFETY: we have to return a known sentinel here that cannot be produced for
+            // a valid pointer, so that `from_raw` can reverse this transformation.
+            (ptr as *mut T).set_ptr_value(ptr::null_mut())
+        } else {
+            // SAFETY: If the pointer is not dangling, it describes to a valid allocation.
```

Comment:
"it describes to a valid allocation" doesn't sound right; not entirely sure what you mean though.

Reply:
I said both "describes a valid allocation" and "points to a valid allocation" at the same time, whoops 🙃

Points to a known-valid allocation (i.e. deref (to `MaybeUninit`) is fine), but the `T` behind the pointer may be in a dropped state.

Reply:
```suggestion
            // SAFETY: If the pointer is not dangling, it describes a valid allocation.
```

Reply:
> Points to a known-valid allocation (i.e. deref (to MaybeUninit) is fine), but the T behind the pointer may be in a dropped state.

Yeah, something like that, or maybe "is dereferencable".

I don't think I saw the term "describe" used for a pointer before.

Reply:
Interestingly enough, "describe" is used by `data_offset`'s docs here.

I personally have used "pointer describes" to include (address, metadata), and "pointer points" to include (address, provenance). This is definitely not standard/widespread, though, so a different wording is probably better. I'd rather not the comment be "if the pointer is not dangling, the pointer is not dangling," though 😛

Reply:
>  "if the pointer is not dangling, the pointer is not dangling,"

Well, that comment would at least not be wrong. :P

"If `is_dangling` returns `false`, then the pointer is dereferencable"?

Reply:
I put "If the pointer is not dangling, it references a valid allocation." in the update I just pushed, but I actually like yours slightly better in this case.

Reply:
Just pushed another update which should be satisfactory.

`as_ptr`

```rust
        if is_dangling(self.ptr) {
            // If the pointer is dangling, we return a null pointer as the dangling sentinel.
            [...]
        } else {
            // SAFETY: if is_dangling returns false, then the pointer is dereferencable.
            [...]
```

`from_raw`

        if ptr.is_null() {
            // If we get a null pointer, this is a dangling weak.
            [...]
        } else {
            // Otherwise, we're guaranteed the pointer came from a nondangling Weak.
            [...]
```

which also hopefully makes it clear that the potential optimization for sized types (of using branchless `wrapping_offset` rather than `if dangling`) requires changing `from_raw` (to use `wrapping_offset`) as well.

Reply:
> which also hopefully makes it clear that the potential optimization for sized types (of using branchless wrapping_offset rather than if dangling) requires changing from_raw (to use wrapping_offset) as well.

It doesn't really make it clear to me, but I also have not thought about the optimization, so that's okay.

The text looks good, thanks!